File: scripts/generate_brownfield_seed.py
#!/usr/bin/env python3
"""Parse Capex FY 2026-27 RAC Plants workbook and emit brownfield seed TS."""
from __future__ import annotations

import json
import re
from pathlib import Path

import openpyxl

ROOT = Path(__file__).resolve().parents[1]
XLSX = ROOT / "Capex FY 2026-27 RAC Plants final 18 April.xlsx"
OUT = ROOT / "src/lib/brownFieldSeedData.ts"

SHEET_PLANT = {
    "DDN-4": "ddn_4",
    "DDN-5": "ddn_5",
    "DDN-6": "ddn_6",
    "JJR P1": "jhajjar_p1",
    "JJR P2": "jhajjar_p2",
    "SUPA": "supa",
    "Rudrapur": "rudrapur",
    "Sricity-1": "sircity_1",
    "Sricity-2": "sircity_2",
}

FY = "2026-27"
CR = 10_000_000


def norm_head(h: str | None) -> str:
    if not h:
        return ""
    s = str(h).strip()
    s = re.sub(r", If any$", "", s, flags=re.I)
    mapping = {
        "genral": "General",
        "general": "General",
        "new business": "New Business",
        "misc.": "Misc.",
        "misc": "Misc.",
    }
    low = s.lower()
    for k, v in mapping.items():
        if low == k or low.startswith(k):
            if k == "new business":
                return "New Business"
            return v
    return s


def is_total_row(sno, sub) -> bool:
    s = str(sno).strip().lower() if sno is not None else ""
    sub_s = str(sub).strip().lower() if sub is not None else ""
    if s in ("total",) or s.startswith("total "):
        return True
    if sub_s in ("total", "total value required"):
        return True
    return False


def normalize_rate_rs(rate_val, qty_val, total_cr: float | None) -> int | None:
    if rate_val is None or rate_val == "":
        return None
    try:
        rate = float(rate_val)
    except (TypeError, ValueError):
        return None
    if not (rate == rate):  # NaN
        return None

    qty = None
    if qty_val is not None and qty_val != "":
        try:
            qty = float(qty_val)
        except (TypeError, ValueError):
            qty = None

    if rate > 0 and rate < 100:
        if qty and qty > 0 and total_cr is not None:
            expected = rate * qty
            if abs(expected - total_cr) < 0.02:
                return round((total_cr / qty) * CR)
        if (qty is None or qty == 1) and total_cr is not None and abs(rate - total_cr) < 0.001:
            return round(total_cr * CR)
        return round(rate * CR)
    return round(rate)


def find_header_row(ws):
    for i, row in enumerate(ws.iter_rows(values_only=True), start=1):
        vals = [str(v).strip() if v is not None else "" for v in row]
        if any(v.lower().startswith("s.no") for v in vals):
            return i, list(row)
    return None, None


def col_index(headers, *needles):
    lower = [str(h).lower().replace("\n", " ") if h is not None else "" for h in headers]
    for idx, h in enumerate(lower):
        if all(n in h for n in needles):
            return idx
    return None


def parse_sheet(ws, plant: str) -> list[dict]:
    header_row, headers = find_header_row(ws)
    if not header_row:
        return []

    lower = [str(h).lower().replace("\n", " ") if h is not None else "" for h in headers]
    c_sno = col_index(headers, "s.no")
    c_head = col_index(headers, "head")
    c_sub = col_index(headers, "sub")
    c_dept = col_index(headers, "department") or col_index(headers, "departments")
    c_rate = col_index(headers, "rate")
    c_qty = next((i for i, h in enumerate(lower) if "qty" in h), None)
    tc_cols = [i for i, h in enumerate(lower) if "total cost" in h and "cr" in h]
    c_total = tc_cols[-1] if len(tc_cols) > 1 else (tc_cols[0] if tc_cols else None)
    c_reason = col_index(headers, "reason")
    c_benefits = col_index(headers, "benefit")
    c_roi = col_index(headers, "roi")

    items = []
    current_head = ""

    for row in ws.iter_rows(min_row=header_row + 1, values_only=True):
        row = list(row)

        def get(c):
            return row[c] if c is not None and c < len(row) else None

        sno_raw = get(c_sno)
        sub_raw = get(c_sub)
        head_raw = get(c_head)

        if sno_raw is None and (sub_raw is None or sub_raw == ""):
            continue

        if head_raw not in (None, ""):
            current_head = norm_head(head_raw)

        sub = str(sub_raw).strip() if sub_raw is not None else ""
        sno_str = str(sno_raw).strip() if sno_raw is not None else ""

        total_cr_val = get(c_total)
        total_cost = float(total_cr_val) if total_cr_val is not None and str(total_cr_val) not in ("", "#DIV/0!") else 0.0

        is_uc = plant == "jhajjar_p1" and sub and re.search(r"urban company|^uc$", sub, re.I)

        is_sricity_placeholder = (
            plant == "sircity_2"
            and current_head == "New Business"
            and not sub
            and total_cost > 1
        )
        # Row 39 is a sheet-level total rollup, not an additional budget line.
        if is_sricity_placeholder:
            continue

        if is_total_row(sno_raw, sub_raw) and not is_uc:
            continue
        if not sub and not is_uc:
            continue

        qty_raw = get(c_qty)
        qty = None
        if qty_raw is not None and str(qty_raw).strip() not in ("", "-", "#DIV/0!"):
            try:
                qty = float(qty_raw)
            except (TypeError, ValueError):
                qty = None
        rate_rs = normalize_rate_rs(get(c_rate), qty_raw, total_cost)
        rate_cr = (rate_rs / CR) if rate_rs is not None else 0.0

        reason = get(c_reason)
        benefits = get(c_benefits)
        roi_raw = get(c_roi)
        roi = None
        if roi_raw is not None and str(roi_raw).strip() not in ("", "-", "#DIV/0!"):
            roi = str(roi_raw).strip()

        dept_raw = get(c_dept)
        department = str(dept_raw).strip() if dept_raw is not None else ""

        item = {
            "plant": plant,
            "head": current_head or "General",
            "department": department,
            "division": "Other Brown Field",
            "subParticulars": sub
            if not is_uc
            else "UC (Urban Company — already approved Nov 2025, balance shifted to FY26-27)",
            "rate": rate_cr,
            "totalCost": round(total_cost, 10),
            "fieldType": "brown_field",
            "fy": FY,
        }
        if sno_str and not is_uc:
            item["sNo"] = sno_str
        if is_uc:
            item["sNo"] = "NB1"
        if rate_rs is not None:
            item["rateRs"] = rate_rs
        if qty is not None:
            item["qty"] = qty
        if reason:
            item["reasonForRequirement"] = str(reason).strip()
        if benefits:
            item["benefits"] = str(benefits).strip()
        if roi:
            item["roi"] = roi

        items.append(item)

    return items


def esc_ts(s: str) -> str:
    return json.dumps(s, ensure_ascii=False)


def main():
    wb = openpyxl.load_workbook(XLSX, data_only=True, read_only=True)
    all_items: list[dict] = []

    for sheet_name, plant in SHEET_PLANT.items():
        ws = wb[sheet_name]
        rows = parse_sheet(ws, plant)
        all_items.extend(rows)
        total = sum(r["totalCost"] for r in rows)
        print(f"{sheet_name} ({plant}): {len(rows)} rows, {total:.3f} Cr")

    totals: dict[str, float] = {}
    for item in all_items:
        totals[item["plant"]] = totals.get(item["plant"], 0) + item["totalCost"]
    print("\nPlant totals:", {k: round(v, 3) for k, v in totals.items()})
    print("Grand total:", round(sum(totals.values()), 3))

    lines = [
        "import type { CapexMasterItem } from './types';",
        "",
        "/** Bump when Brown Field workbook seed is regenerated — triggers one-time localStorage migration. */",
        "export const BROWNFIELD_SEED_VERSION = 'fy2026_27_rac_plants';",
        "",
        "/** FY 2026-27 Brown Field RAC plant master — generated from workbook. Do not edit by hand. */",
        "export const brownFieldSeedData: CapexMasterItem[] = [",
    ]

    for idx, item in enumerate(all_items):
        item_id = f"cm-bf-{item['plant']}-{idx + 1:04d}"
        parts = [
            f"id: {esc_ts(item_id)}",
            "fieldType: 'brown_field'",
            f"fy: {esc_ts(FY)}",
            f"plant: {esc_ts(item['plant'])}",
            f"division: {esc_ts(item.get('division', 'Other Brown Field'))}",
            f"head: {esc_ts(item['head'])}",
            f"department: {esc_ts(item.get('department', ''))}",
            f"subParticulars: {esc_ts(item['subParticulars'])}",
            f"rate: {item['rate']}",
            f"totalCost: {item['totalCost']}",
        ]
        if item.get("sNo"):
            parts.append(f"sNo: {esc_ts(item['sNo'])}")
        if item.get("rateRs") is not None:
            parts.append(f"rateRs: {item['rateRs']}")
        if item.get("qty") is not None:
            parts.append(f"qty: {item['qty']}")
        if item.get("reasonForRequirement"):
            parts.append(f"reasonForRequirement: {esc_ts(item['reasonForRequirement'])}")
        if item.get("benefits"):
            parts.append(f"benefits: {esc_ts(item['benefits'])}")
        if item.get("roi"):
            parts.append(f"roi: {esc_ts(item['roi'])}")

        lines.append(f"  {{ {', '.join(parts)} }},")

    lines.append("];")
    lines.append("")
    OUT.write_text("\n".join(lines), encoding="utf-8")
    print(f"\nWrote {len(all_items)} items to {OUT}")


if __name__ == "__main__":
    main()

File: scripts/generate-brownfield-seed.mjs
/**
 * Parses Capex FY 2026-27 RAC Plants workbook and emits brownfield seed TS.
 * Run: node scripts/generate-brownfield-seed.mjs
 */
import ExcelJS from 'exceljs';
import { writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');
const XLSX = join(ROOT, 'Capex FY 2026-27 RAC Plants final 18 April.xlsx');
const OUT = join(ROOT, 'src/lib/brownFieldSeedData.ts');

const SHEET_PLANT = {
  'DDN-4': 'ddn_4',
  'DDN-5': 'ddn_5',
  'DDN-6': 'ddn_6',
  'JJR P1': 'jhajjar_p1',
  'JJR P2': 'jhajjar_p2',
  SUPA: 'supa',
  Rudrapur: 'rudrapur',
  'Sricity-1': 'sircity_1',
  'Sricity-2': 'sircity_2',
};

const FY = '2026-27';
const CR = 10_000_000;

function normHead(h) {
  if (!h) return '';
  let s = String(h).trim();
  s = s.replace(/, If any$/i, '').replace(/, if any$/i, '');
  if (/^genral$/i.test(s) || /^general$/i.test(s)) return 'General';
  if (/^new business/i.test(s)) return 'New Business';
  if (/^misc\.?$/i.test(s)) return 'Misc.';
  if (/^safety$/i.test(s)) return 'Safety';
  if (/^quality$/i.test(s)) return 'Quality';
  if (/^compliance$/i.test(s)) return 'Compliance';
  if (/^moulding merger$/i.test(s)) return 'Moulding Merger';
  if (/^productivity improvement$/i.test(s)) return 'Productivity Improvement';
  if (/^standardization$/i.test(s)) return 'Standardization';
  if (/^business requirement$/i.test(s)) return 'Business Requirement';
  return s;
}

function isTotalRow(sno, sub, head) {
  const s = sno != null ? String(sno).trim().toLowerCase() : '';
  const subS = sub != null ? String(sub).trim().toLowerCase() : '';
  if (s === 'total' || s.startsWith('total ')) return true;
  if (subS === 'total' || subS === 'total value required') return true;
  return false;
}

function normalizeRateRs(rateVal, qtyVal, totalCr) {
  if (rateVal == null || rateVal === '') return null;
  const rate = Number(rateVal);
  if (!Number.isFinite(rate)) return null;
  const qty = qtyVal != null && qtyVal !== '' ? Number(qtyVal) : null;

  // Crore-like rate (JJR P1/P2, DDN-5): value < 100 and totalCr matches rate*qty when qty present
  if (rate > 0 && rate < 100) {
    if (qty != null && Number.isFinite(qty) && qty > 0 && totalCr != null) {
      const expectedCr = rate * qty;
      if (Math.abs(expectedCr - totalCr) < 0.02) {
        return Math.round((totalCr / qty) * CR);
      }
    }
    // Single qty crore rate
    if ((qty == null || qty === 1) && totalCr != null && Math.abs(rate - totalCr) < 0.001) {
      return Math.round(totalCr * CR);
    }
    return Math.round(rate * CR);
  }
  return Math.round(rate);
}

function esc(s) {
  if (s == null) return "''";
  return `'${String(s).replace(/\\/g, '\\\\').replace(/'/g, "\\'").replace(/\n/g, '\\n')}'`;
}

function findHeaderRow(sheet) {
  for (let r = 1; r <= Math.min(sheet.rowCount, 10); r++) {
    const row = sheet.getRow(r);
    const vals = [];
    row.eachCell({ includeEmpty: true }, (cell, col) => {
      vals[col - 1] = cell.value;
    });
    const lower = vals.map((v) => (v != null ? String(v).toLowerCase().replace(/\n/g, ' ') : ''));
    if (lower.some((h) => h.startsWith('s.no'))) {
      return { rowIndex: r, headers: vals };
    }
  }
  return null;
}

function colIndex(headers, ...needles) {
  const lower = headers.map((h) => (h != null ? String(h).toLowerCase().replace(/\n/g, ' ') : ''));
  for (let i = 0; i < lower.length; i++) {
    if (needles.every((n) => lower[i].includes(n))) return i;
  }
  return -1;
}

async function parseSheet(sheet, plant) {
  const header = findHeaderRow(sheet);
  if (!header) return [];

  const { rowIndex, headers } = header;
  const cSno = colIndex(headers, 's.no');
  const cHead = colIndex(headers, 'head');
  const cSub = colIndex(headers, 'sub');
  const cDept =
    colIndex(headers, 'department') >= 0
      ? colIndex(headers, 'department')
      : colIndex(headers, 'departments');
  const cRate = colIndex(headers, 'rate');
  const cQty = headers.findIndex((h) => h != null && String(h).toLowerCase().includes('qty'));
  const tcCols = headers
    .map((h, i) => ({ h, i }))
    .filter(({ h }) => h != null && String(h).toLowerCase().includes('total cost') && String(h).toLowerCase().includes('cr'))
    .map(({ i }) => i);
  const cTotal = tcCols.length > 1 ? tcCols[tcCols.length - 1] : tcCols[0] ?? -1;
  const cReason = colIndex(headers, 'reason');
  const cBenefits = colIndex(headers, 'benefit');
  const cRoi = colIndex(headers, 'roi');

  const items = [];
  let currentHead = '';

  for (let r = rowIndex + 1; r <= sheet.rowCount; r++) {
    const row = sheet.getRow(r);
    const get = (c) => (c >= 0 ? row.getCell(c + 1).value : null);

    const snoRaw = get(cSno);
    const subRaw = get(cSub);
    const headRaw = get(cHead);

    if (snoRaw == null && (subRaw == null || subRaw === '')) continue;

    if (headRaw != null && String(headRaw).trim() !== '') {
      currentHead = normHead(headRaw);
    }

    const sub = subRaw != null ? String(subRaw).trim() : '';
    const snoStr = snoRaw != null ? String(snoRaw).trim() : '';

    // Include UC and Sricity-2 placeholder even without sub
    const isUc =
      plant === 'jhajjar_p1' &&
      (snoStr.toUpperCase() === 'NB1' || (sub && /urban company|^uc$/i.test(sub)));
    const isSricityPlaceholder =
      plant === 'sircity_2' &&
      currentHead === 'New Business' &&
      !sub &&
      get(cTotal) != null &&
      Number(get(cTotal)) > 1;

    if (isTotalRow(snoRaw, subRaw, currentHead) && !isUc && !isSricityPlaceholder) continue;

    if (!sub && !isUc && !isSricityPlaceholder) continue;

    const totalCrVal = get(cTotal);
    const totalCost =
      totalCrVal != null && totalCrVal !== '' && Number.isFinite(Number(totalCrVal))
        ? Number(totalCrVal)
        : 0;

    const qtyRaw = get(cQty);
    const qty =
      qtyRaw != null && qtyRaw !== '' && Number.isFinite(Number(qtyRaw)) ? Number(qtyRaw) : undefined;

    const rateRs = normalizeRateRs(get(cRate), qtyRaw, totalCost);
    const rateCr = rateRs != null ? rateRs / CR : 0;

    const reason = get(cReason);
    const benefits = get(cBenefits);
    const roiRaw = get(cRoi);
    const roi =
      roiRaw == null || roiRaw === '' || String(roiRaw).trim() === '-'
        ? undefined
        : String(roiRaw).trim();

    const deptRaw = get(cDept);
    const department = deptRaw != null ? String(deptRaw).trim() : '';

    const sNo = snoStr || undefined;

    items.push({
      plant,
      sNo,
      head: currentHead || 'General',
      department,
      subParticulars: sub || (isSricityPlaceholder ? 'New business items (FY 2026-27)' : 'UC (Urban Company)'),
      rateRs,
      qty,
      totalCost,
      rate: rateCr,
      reasonForRequirement: reason != null ? String(reason).trim() : undefined,
      benefits: benefits != null ? String(benefits).trim() : undefined,
      roi,
    });
  }

  return items;
}

async function main() {
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.readFile(XLSX);

  const all = [];
  for (const [sheetName, plant] of Object.entries(SHEET_PLANT)) {
    const sheet = wb.getWorksheet(sheetName);
    if (!sheet) {
      console.warn('Missing sheet:', sheetName);
      continue;
    }
    const rows = await parseSheet(sheet, plant);
    all.push(...rows);
    const sum = rows.reduce((s, i) => s + i.totalCost, 0);
    console.log(`${sheetName} (${plant}): ${rows.length} rows, ${sum.toFixed(3)} Cr`);
  }

  const totals = {};
  for (const item of all) {
    totals[item.plant] = (totals[item.plant] ?? 0) + item.totalCost;
  }
  console.log('\nPlant totals:', totals);
  console.log('Grand total:', Object.values(totals).reduce((a, b) => a + b, 0).toFixed(3));

  const lines = [
    "import type { CapexMasterItem } from './types';",
    '',
    '/** FY 2026-27 Brown Field RAC plant master — generated from workbook. Do not edit by hand. */',
    'export const brownFieldSeedData: CapexMasterItem[] = [',
  ];

  all.forEach((item, idx) => {
    const id = `cm-bf-${item.plant}-${String(idx + 1).padStart(4, '0')}`;
    const parts = [
      `id: ${esc(id)}`,
      `fieldType: 'brown_field'`,
      `fy: ${esc(FY)}`,
      `plant: ${esc(item.plant)}`,
      `head: ${esc(item.head)}`,
      `department: ${esc(item.department)}`,
      `subParticulars: ${esc(item.subParticulars)}`,
      `rate: ${item.rate}`,
      `totalCost: ${item.totalCost}`,
    ];
    if (item.sNo) parts.push(`sNo: ${esc(item.sNo)}`);
    if (item.rateRs != null) parts.push(`rateRs: ${item.rateRs}`);
    if (item.qty != null) parts.push(`qty: ${item.qty}`);
    if (item.reasonForRequirement) parts.push(`reasonForRequirement: ${esc(item.reasonForRequirement)}`);
    if (item.benefits) parts.push(`benefits: ${esc(item.benefits)}`);
    if (item.roi) parts.push(`roi: ${esc(item.roi)}`);

    lines.push(`  { ${parts.join(', ')} },`);
  });

  lines.push('];', '');
  writeFileSync(OUT, lines.join('\n'));
  console.log(`\nWrote ${all.length} items to ${OUT}`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
